src/context.cpp

```diff
@@ -38,42 +42,168 @@
+    // because services are allowed to start and stop other services during their lifetime.
+    synchronized<service_list_t> services;
+
+    // Context signalling hub.
+    retroactive_signal<io::context_tag> signals;
 
-    const holder_t scoped(*m_log, {{"source", "core"}});
+    const config_t config;
 
-    COCAINE_LOG_INFO(m_log, "initializing the core");
+    // Service port mapping and pinning.
+    port_mapping_t mapper;
 
-    m_repository = std::make_unique<api::repository_t>(log("repository"));
+};
+
+void
+context_t::impl_deleter_t::operator()(impl_t* ptr) const {
+    delete ptr;
+}
+
+
+context_t::context_t(config_t config, std::unique_ptr<logging::logger_t> logger) :
+    impl(new impl_t(std::move(config), std::move(logger))) {
+
+    const holder_t scoped(*impl->log, {{"source", "core"}});
+
+    COCAINE_LOG_INFO(impl->log, "initializing the core");
+
+    impl->repository = std::make_unique<api::repository_t>(log("repository"));
 
     // Load the builtin plugins.
-    essentials::initialize(*m_repository);
+    essentials::initialize(*impl->repository);
 
     // Load the rest of plugins.
-    m_repository->load(config.path.plugins);
+    impl->repository->load(impl->config.path.plugins);
 
     // Spin up all the configured services, launch execution units.
-    bootstrap();
+    COCAINE_LOG_INFO(impl->log, "starting {:d} execution unit(s)", config.network.pool);
+
+    while (impl->pool.size() != config.network.pool) {
+        impl->pool.emplace_back(std::make_unique<execution_unit_t>(*this));
+    }
+
+    COCAINE_LOG_INFO(impl->log, "starting {:d} service(s)", config.services.size());
+
+    std::vector<std::string> errored;
+
+    for (auto it = config.services.begin(); it != config.services.end(); ++it) {
+        const holder_t scoped(*impl->log, {{"service", it->first}});
+
+        const auto asio = std::make_shared<asio::io_service>();
+
+        COCAINE_LOG_DEBUG(impl->log, "starting service");
+
+        try {
+            insert(it->first, std::make_unique<actor_t>(*this, asio, repository().get<api::service_t>(
+            it->second.type,
+            *this,
+            *asio,
+            it->first,
+            it->second.args
+            )));
+        } catch (const std::system_error& e) {
+            COCAINE_LOG_ERROR(impl->log, "unable to initialize service: {}", error::to_string(e));
+            errored.push_back(it->first);
+        } catch (const std::exception& e) {
+            COCAINE_LOG_ERROR(impl->log, "unable to initialize service: {}", e.what());
+            errored.push_back(it->first);
+        }
+    }
+
+    if (!errored.empty()) {
+        COCAINE_LOG_ERROR(impl->log, "emergency core shutdown");
+
+        // Signal and stop all the services, shut down execution units.
+        terminate();
+
+        std::ostringstream stream;
+        std::ostream_iterator<char> builder(stream);
+
+        boost::spirit::karma::generate(builder, boost::spirit::karma::string % ", ", errored);
```

Comment:
I think we should drop karma to the Hell here and elsewhere. But it's out of the scope of this PR.